Extend and cleanup usability of __CPROVER_{r,w}_ok

tautschnig opened the pull request:
`__CPROVER_r_ok` and `__CPROVER_w_ok` had surprising behaviour that wasn't documented. Make the behaviour less surprising and document it.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Always evaluate __CPROVER_{r,w}_ok and add a test

We did not have an explicit test of __CPROVER_{r,w}_ok, only implicit
ones as we use these predicates in our model of the C library. While
preparing a test it became apparent that the predicates are only
evaluated when --pointer-check is set. This caused surprising behaviour
as a negated predicate would result in failing assertions. Instead, make
sure the expression is always evaluated, independent of --pointer-check.
This requires always performing a local alias analysis, even when no
pointer checks are enabled.

danielsn commented:
You might want to add a test where "r_ok" should pass, but "w_ok" should fail. Something like
```
void test() {
  const char* str = "foobar";
  assert(!__CPROVER_w_ok(str, 6));
  assert(__CPROVER_r_ok(str, 6));
}
```

tautschnig committed:
Add test showing limitations of __CPROVER_w_ok

We do not currently have a good way of distinguishing lvalues from
rvalues, and thus actually treat __CPROVER_w_ok and __CPROVER_r_ok the
same. The test shows that this shouldn't always be done.

tautschnig commented:
> You might want to add a test where "r_ok" should pass, but "w_ok" should fail.

Thank you very much for the suggestion - I have added such a test, but marked it as `KNOWNBUG` as we do not currently have a way do distinguish lvalues at this point.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7f8f20e6d13accf56faaa6badac9ccbccec8fe00).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107129865
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

danielsn commented on regression/cbmc/r_w_ok1/main.c:
```diff
@@ -0,0 +1,26 @@
+  assert(__CPROVER_r_ok(arbitrary_size, n));
+  assert(__CPROVER_w_ok(arbitrary_size, n));
+
+  assert(__CPROVER_r_ok(arbitrary_size, n + 1));
```
Why does this succeed?

tautschnig replied:
Without `--pointer-check` any use of `__CPROVER_{r,w}_ok` is now a no-op (and is now also documented as such), but with `--pointer-check` this fails as expected (there are two regression tests using the same `main.c` file included).

kroening replied:
Is this what we want? I would find it more intuitive if these were independent.

tautschnig replied:
It does make the implementation a lot easier, and also removes a need for documentation, so that seems much better. The only cost is that we will unconditionally need to run the local alias analysis.

tautschnig replied:
Ok, this will require #4471 to be merged first. And another bug left to be debugged.

tautschnig replied:
Hmm, actually there is more to it than just #4471, we also seem to be messing up `ID_this` elsewhere.

kroening replied:
If the cost of the local alias analysis is a concern, there could be a quick scan for the predicate in the function; but so far, that analysis has not been cause for concern.

tautschnig replied:
Agreed, I just need to fix bugs in the C++ and Java front-ends first as the local alias analysis needs to succeed in doing symbol-table lookups...

tautschnig committed:
Support __CPROVER_{r,w}_ok in assignments, assumptions, gotos, returns

We previously only evaluated __CPROVER_{r,w}_ok in assertions (without
telling anyone that was the case). Instead, evaluate it in all contexts
where it might reasonably appear.

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -1765,6 +1773,15 @@ void goto_checkt::goto_check(
       // the LHS might invalidate any assertion
       invalidate(code_assign.lhs());
+
+      if(has_subexpr(i.code.op1(), [](const exprt &expr) {
```
`i.code.op1` -> `code_assign.rhs`

tautschnig replied:
Done.

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -1765,6 +1773,15 @@ void goto_checkt::goto_check(
+      {
+        auto rw_ok_cond = rw_ok_check(i.code.op1());
+        if(rw_ok_cond.has_value())
+          i.code.op1() = *rw_ok_cond;
```
Hang on, so if any subexpr was an r_ok or w_ok then you overwrite the whole RHS? That seems surprising, what's the rationale for blowing away arbitrary AST nodes here?

tautschnig replied:
No, `rw_ok_check` will take care of only replacing parts of the RHS.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1a8b8340d1c5f7a31af0710ea59c5d28cba2454f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107785716
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -1813,6 +1831,15 @@ void goto_checkt::goto_check(
+             return expr.id() == ID_r_ok || expr.id() == ID_w_ok;
+           }))
+        {
+          auto rw_ok_cond = rw_ok_check(i.code.op0());
```
`to_code_return(code).return_value()`

tautschnig replied:
Oops, thank you for keeping me from being lazy.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d7dc902cc535ac0115b1bd765182f865cc15fa13).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107793523
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7ecf1f69e859f76b5aa5ae86c2ff2ba9ccae62ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107802062
